src/scene/materials/material-registry.js

```diff
@@ -0,0 +1,55 @@
+import { Shader } from '../../graphics/shader.js';
+
+class MaterialRegistry {
+    constructor() {
+        this.materials = new Set();
+        Shader.debugEvents.on('compileError', (shader) => this.reportEffectedMaterials(shader));
+        Shader.debugEvents.on('linkError', (shader) => this.reportEffectedMaterials(shader));
+    }
+
+    add(material) {
+        this.materials.add(material);
+    }
+
+    remove(material) {
+        this.materials.delete(material);
+    }
+
+    reportEffectedMaterials(shader) {
+        // search for all materials making use of this shader
+        const materials = new Set();
+        this.materials.forEach((material) => {
+            // check the shader
+            if (material.shader === shader) {
+                materials.add(material);
+            }
+
+            // check material variants
+            Object.keys(material.variant || {}).forEach((variantIndex) => {
+                const variant = material.variant[variantIndex];
+                if (variant === shader) {
+                    materials.add(material);
+                }
+            });
+
+            // check mesh instance variants
+            material.meshInstances.forEach((meshInstance) => {
+                meshInstance._shader.forEach((variant) => {
+                    if (variant === shader) {
+                        materials.add(material);
+                    }
+                });
+            });
+        });
+
+        if (materials.size > 0) {
+            console.error(`Effected materials: `, Array.from(materials));
+        }
+    }
+}
+
+const materialRegistry = new MaterialRegistry();
```

Comment:
So all running applications share the same registry? 😨 Are we cool with that?

Reply:
I'm not. If an application won't destroy some materials, they will stay in memory as the registry would keep the references.

Reply:
How would shader know which application it belongs to?

Reply:
I think it needs to be per device and not per application, like the other device things we store using DeviceCache.

Reply:
The issue is materials are not application or device specific.

Reply:
Ideally the registry would be per-application, but I don't see how.

But anyway I don't think the registry hanging onto un-destroyed materials is such a big deal because:
- they're not very big (compared to texture data for example)
- this is only a potential issue in debug builds running multiple applications
- if it ever was an issue, users could just delete their materials

And actually having access to the list of undeleted materials could be extremely useful to users, something we might want to expose in future.

Reply:
they are valid for use only on a single device, so they are device specific. For example shader and textures they reference are all device specific, making them the same.
But which device is not known ahead of time. Perhaps at a point the shader is assigned to the material (which has a device), the material could be added to the cache, and not in the constructor? It could even be a Set on a shader directly?